File: packages/ogt-py/src/ogt/draw/screws.py
"""Naive screw hole cutout geometry."""

import math

import cadquery as cq

from ogt.prepare.types import ScrewSize

SCREW_HEAD_COUNTERSUNK_DEGREE = 90


def make_screw_cutout(
    screw_size: ScrewSize = ScrewSize(),
    tile_thickness: float = 6.8,
    head_at_bottom: bool = False,
) -> cq.Workplane:
    """Build a single screw hole cutout tool.

    Three-part geometry:
    1. Main cylinder through the full tile thickness.
    2. Countersink cone.
    3. Head (counterbore) cylinder.

    When *head_at_bottom* is False (default), the head sits at the top
    of the tile (full tile orientation).  When True, the head sits at
    the bottom (lite tile orientation).
    """
    main_r = screw_size.diameter / 2
    head_r = screw_size.head_diameter / 2
    countersink_h = math.tan(math.radians(SCREW_HEAD_COUNTERSUNK_DEGREE / 2)) * (head_r - main_r)

    # Main cylinder, full height through tile
    main_cyl = cq.Workplane("XY").circle(main_r).extrude(tile_thickness)

    if head_at_bottom:
        # Head (counterbore) at Z=0
        head = cq.Workplane("XY").circle(head_r).extrude(screw_size.head_inset)

        # Countersink cone above the counterbore (head_r -> main_r going up)
        countersink = (
            cq.Workplane("XY")
            .circle(head_r)
            .workplane(offset=countersink_h)
            .circle(main_r)
            .loft()
            .translate((0, 0, screw_size.head_inset))
        )
    else:
        # Countersink cone near the top (main_r -> head_r going up)
        countersink_base_z = tile_thickness - screw_size.head_inset - countersink_h
        countersink = (
            cq.Workplane("XY")
            .circle(main_r)
            .workplane(offset=countersink_h)
            .circle(head_r)
            .loft()
            .translate((0, 0, countersink_base_z))
        )

        # Head inset cylinder at the top
        head = (
            cq.Workplane("XY")
            .circle(head_r)
            .extrude(screw_size.head_inset)
            .translate((0, 0, tile_thickness - screw_size.head_inset))
        )

    return main_cyl.union(countersink).union(head)

File: packages/ogt-web/src/lib/cad/screws.ts
import { drawCircle, makeCylinder, type Solid, type Sketch } from "replicad";
import { TILE_THICKNESS } from "./constants";
import type { ScrewSize } from "../types";

const SCREW_HEAD_COUNTERSUNK_DEGREE = 90;

const DEFAULT_SCREW: ScrewSize = {
  diameter: 4.2,
  head_diameter: 8.0,
  head_inset: 1.0,
};

export function makeScrewCutout(screwSize: ScrewSize = DEFAULT_SCREW): Solid {
  const mainR = screwSize.diameter / 2;
  const headR = screwSize.head_diameter / 2;
  const countersinkH =
    Math.tan((SCREW_HEAD_COUNTERSUNK_DEGREE / 2) * (Math.PI / 180)) *
    (headR - mainR);

  // Main cylinder, full height through tile
  const mainCyl = makeCylinder(mainR, TILE_THICKNESS, [0, 0, 0], [0, 0, 1]);

  // Countersink cone via loft
  const countersinkBaseZ = TILE_THICKNESS - screwSize.head_inset - countersinkH;
  const bottomCircle = drawCircle(mainR).sketchOnPlane(
    "XY",
    countersinkBaseZ,
  ) as Sketch;
  const topCircle = drawCircle(headR).sketchOnPlane(
    "XY",
    countersinkBaseZ + countersinkH,
  ) as Sketch;
  const countersink = bottomCircle.loftWith(topCircle) as Solid;

  // Head inset cylinder
  const head = makeCylinder(
    headR,
    screwSize.head_inset,
    [0, 0, TILE_THICKNESS - screwSize.head_inset],
    [0, 0, 1],
  );

  return mainCyl.fuse(countersink).fuse(head) as Solid;
}
